File: frontend/src/types/apply.ts
/**
 * 「自动采集 + 匹配度分析 + 自动投递」的前端类型镜像。
 *
 * 这里的字符串字面量联合类型与 `backend/app/models/apply.py` 的常量**逐字一致**，
 * 改一处必须同步另一处（设计 §9 ⑨）。界面只读后端返回的 `admission` / `requires_confirm`，
 * **不在这里再判一次准入**——判断的权威只有后端 `admission_of()` 一处。
 */

// ===== ① 匹配状态（五类）=====
export type MatchStatus =
  "matched" | "expression_gap" | "evidence_insufficient" | "real_gap" | "to_confirm";

export type HardGateResult = "met" | "unmet" | "unknown";
export type AdmissionResult = "allow" | "block" | "needs_confirm";

export interface MatchCondition {
  label: string;
  jd_quote: string;
  status: MatchStatus;
  evidence: string;
}

export interface JobMatchResult {
  hard_conditions: MatchCondition[];
  core_abilities: MatchCondition[];
  bonus_items: MatchCondition[];
  hard_gate: HardGateResult;
  admission: AdmissionResult;
  advice: string;
  notes: string[];
}

/** 参考分的单个分项（0-100 分 + 权重 + 一句可读的依据）。 */
export interface MatchScoreDimension {
  key: string;
  label: string;
  score: number;
  weight: number;
  evidence: string;
}

/** 匹配度参考分：0-100 总分 + 5 个分项 + **后端下发的免责文案**。
 *
 * 它是**纯本地规则**算出来的派生值（不调模型）、**不参与投递准入**——能不能投仍只看
 * 五类结论。`disclaimer` 必须原样展示，不要自己改写或省略。
 */
export interface MatchReferenceScore {
  score: number;
  dimensions: MatchScoreDimension[];
  disclaimer: string;
}

export interface JobMatchOut {
  id: number;
  job_id: number | null;
  job_title: string;
  company: string;
  result: JobMatchResult;
  hard_gate: HardGateResult;
  requires_confirm: boolean;
  model: string;
  created_at: string;
  updated_at: string;
  /** 派生值、仅展示、带免责；未分析过时为 null。 */
  reference_score: MatchReferenceScore | null;
}

/** 五类状态的中文名与展示色；`admission` 与后端闸门映射一致，仅用于说明，不用于判定。 */
export const MATCH_STATUS_META: Record<
  MatchStatus,
  { label: string; color: string; admission: AdmissionResult }
> = {
  matched: { label: "已匹配", color: "green", admission: "allow" },
  expression_gap: { label: "表达缺口", color: "blue", admission: "allow" },
  evidence_insufficient: { label: "证据不足", color: "orange", admission: "needs_confirm" },
  real_gap: { label: "真实缺口", color: "red", admission: "block" },
  to_confirm: { label: "待确认", color: "gold", admission: "needs_confirm" },
};

export const ADMISSION_META: Record<AdmissionResult, { label: string; color: string }> = {
  allow: { label: "可投递", color: "green" },
  needs_confirm: { label: "需确认", color: "orange" },
  block: { label: "不投", color: "red" },
};

// ===== ② 任务状态机 =====
export type TaskKind = "collect" | "apply";
export type TaskStatus =
  "pending" | "running" | "paused" | "breaker_paused" | "completed" | "stopped" | "failed";
export type TaskItemStatus = "pending" | "running" | "success" | "failed" | "skipped";
export type QueueStatus = "pending" | "skipped" | "done";
export type BrowserState = "stopped" | "starting" | "running" | "unknown";

export const TASK_STATUS_META: Record<TaskStatus, { label: string; color: string }> = {
  pending: { label: "等待中", color: "default" },
  running: { label: "进行中", color: "processing" },
  paused: { label: "已暂停", color: "warning" },
  breaker_paused: { label: "熔断暂停", color: "error" },
  completed: { label: "已完成", color: "success" },
  stopped: { label: "已停止", color: "default" },
  failed: { label: "失败", color: "error" },
};

export const TASK_ITEM_STATUS_META: Record<TaskItemStatus, { label: string; color: string }> = {
  pending: { label: "待投递", color: "default" },
  running: { label: "进行中", color: "processing" },
  success: { label: "成功", color: "success" },
  failed: { label: "失败", color: "error" },
  skipped: { label: "已跳过", color: "default" },
};

export const QUEUE_STATUS_META: Record<QueueStatus, { label: string; color: string }> = {
  pending: { label: "待投递", color: "default" },
  skipped: { label: "已跳过", color: "default" },
  done: { label: "已投递", color: "success" },
};

export const BROWSER_STATE_META: Record<BrowserState, { label: string; color: string }> = {
  stopped: { label: "未启动", color: "default" },
  starting: { label: "启动中", color: "processing" },
  running: { label: "运行中", color: "success" },
  unknown: { label: "状态未知", color: "warning" },
};

// ===== ③ 失败分类（与后端 FAILURE_CATEGORY_LABELS 一致）=====
export const FAILURE_CATEGORIES = [
  "selector_invalid",
  "login_required",
  "captcha_required",
  "greeting_missing",
  "network_timeout",
  "file_upload_failed",
  "site_unsupported",
  "unknown",
] as const;

export type FailureCategory = (typeof FAILURE_CATEGORIES)[number];

export const FAILURE_CATEGORY_LABELS: Record<FailureCategory, string> = {
  selector_invalid: "页面结构变化 / 选择器失效",
  login_required: "需要登录",
  captcha_required: "需要验证码或安全验证",
  greeting_missing: "招呼语缺失",
  network_timeout: "网络超时",
  file_upload_failed: "简历上传失败",
  site_unsupported: "岗位来源不支持自动投递",
  unknown: "未知失败",
};

/** 后端未给出中文说明时（空串 / 未识别值）回退到原始码或占位。 */
export function failureLabel(category: string, fallback = ""): string {
  if (!category) return fallback;
  return FAILURE_CATEGORY_LABELS[category as FailureCategory] ?? category;
}

// ===== ④ 执行步骤 =====
export const STEP_LABELS: Record<string, string> = {
  opening: "打开岗位页",
  filling: "填写表单",
  greeting: "发送招呼语",
  uploading: "上传简历",
  submitting: "提交投递",
  verifying: "确认结果",
  waiting: "岗位间间隔",
  idle: "空闲",
};

export function stepLabel(step: string): string {
  if (!step) return "空闲";
  return STEP_LABELS[step] ?? step;
}

// ===== ⑤ 配置 =====

/**
 * 投递专用浏览器的选择方式。
 * `auto` = 自动（优先 Chrome，未装回退 Edge）；`custom` = 用下面填的自定义路径。
 */
export type BrowserChoice = "auto" | "chrome" | "edge" | "custom";

export const BROWSER_CHOICE_META: Record<BrowserChoice, { label: string; hint: string }> = {
  auto: { label: "自动（优先 Chrome）", hint: "优先用 Chrome；没装 Chrome 时自动回退 Edge。" },
  chrome: { label: "Google Chrome", hint: "只用 Chrome；找不到会明确提示，不会改用别的浏览器。" },
  edge: { label: "Microsoft Edge", hint: "只用 Edge；找不到会明确提示，不会改用别的浏览器。" },
  custom: {
    label: "自定义路径",
    hint: "填写浏览器可执行文件的完整路径（必须真实存在）。",
  },
};

export interface ApplyConfig {
  interval_seconds: number;
  interval_jitter_seconds: number;
  daily_limit: number;
  per_task_limit: number;
  breaker_threshold: number;
  default_greeting: string;
  skip_same_company: boolean;
  confirm_real_gap: boolean;
  browser_port: number;
  /** 投递台使用的浏览器：自动 / Chrome / Edge / 自定义路径。 */
  browser_choice: BrowserChoice;
  /** 自定义浏览器可执行文件的绝对路径，仅 browser_choice=custom 时生效。 */
  browser_path: string;
  /** 当前对接的招聘网站（站点适配器 key），默认取后端注册表里的第一个站点。 */
  site_key: string;
}

export interface ApplyConfigOut extends ApplyConfig {
  defaults: ApplyConfig;
}

export interface CollectConfig {
  keywords: string[];
  city: string;
  salary_min: number | null;
  experience: string;
  education: string;
  per_task_limit: number;
  interval_seconds: number;
  interval_jitter_seconds: number;
  /** 采集结果标注类型（校招/实习/社招）；可空=不限。只入库标注，不参与站点筛选与去重。 */
  job_type: string;
  /**
   * **站点侧筛选项**：`{分组 key: 选项编码}`（如 `{ degree: "203" }`）。
   *
   * 与上面 `salary_min` / `experience` / `education` 的分工：那三个筛的是"你的条件 vs
   * 岗位要求"（"我是本科"），这一份是**招聘网站筛选栏本身**（"岗位要求本科"）。
   * 选项清单由后端从站点读来（见 `CollectFilterOptions`），这里只存用户选中的编码。
   */
  filters: Record<string, string>;
}

export interface CollectConfigOut extends CollectConfig {
  defaults: CollectConfig;
}

/** 筛选项清单的来源：登录会话 / 全网通用 / 内置快照 / 读不到。 */
export type CollectFilterSource = "session" | "public" | "snapshot" | "unavailable";

export interface CollectFilterOption {
  code: string;
  label: string;
  /** 只用于界面分组（行业有 15 个一级分组），其余为空。 */
  group: string;
}

export interface CollectFilterGroup {
  key: string;
  /** 拼进搜索地址的参数名（由后端实测站点得到，前端不该自己拼）。 */
  param: string;
  label: string;
  options: CollectFilterOption[];
  source: CollectFilterSource;
  note: string;
}

export interface CollectFilterOptions {
  site_key: string;
  display_name: string;
  groups: CollectFilterGroup[];
  /** 是否读到了"你这个登录账号可见"的清单；false 表示用的是全网通用清单。 */
  session_read: boolean;
}

// ===== ⑥ 浏览器状态 =====
export interface BrowserStatus {
  state: BrowserState;
  port: number;
  profile_dir: string;
  browser_path: string;
  /** 人类可读的浏览器名（Google Chrome / Microsoft Edge / 自定义浏览器）。 */
  browser_name: string;
  /** 启动浏览器时会打开的站点入口地址，也用于「打开招聘网站」按钮。 */
  entry_url: string;
  logged_in_hint: string;
  /**
   * 这个浏览器是不是**本次运行**启动的。
   *
   * 为 false 表示它是上一次运行时打开的窗口：`state` 照样是 `running`（调试端口在答，
   * 采集与投递都能用），但「关闭浏览器」关不掉它——进程句柄随后端重启丢了，而应用
   * 只关自己拉起的进程，绝不按 PID 去猜。
   */
  owned: boolean;
}

// ===== ⑩ 招聘网站（当前站点）=====

/**
 * 一个已注册的招聘网站。
 *
 * **前端不写死站点清单**：这里的字段全部来自后端 `GET /api/apply/sites`。以后后端注册表里
 * 多加一个站点，界面自动跟着多出一个选项，前端一行都不用改。
 */
export interface SiteOption {
  key: string;
  display_name: string;
  host: string;
  entry_url: string;
  supports_collect: boolean;
  supports_apply: boolean;
}

export interface SiteList {
  /** 当前选中的站点 key（对应 ApplyConfig.site_key）。 */
  current: string;
  sites: SiteOption[];
}

/** 按 key 找站点展示名；找不到时回退 key 本身，绝不显示空串。 */
export function siteDisplayName(list: SiteList | undefined, key: string): string {
  const found = list?.sites.find((site) => site.key === key);
  return found?.display_name ?? key;
}

// ===== ⑦ 队列 =====
export interface ApplyQueueItem {
  id: number;
  job_id: number | null;
  job_title: string;
  company: string;
  resume_id: number | null;
  resume_title: string;
  greeting: string;
  sort_order: number;
  status: QueueStatus;
  admission: AdmissionResult | null;
  hard_gate: HardGateResult | null;
  requires_confirm: boolean;
  /**
   * 这个岗位能不能自动投递：取决于**来源**是否能归属到某个招聘网站，与匹配结论无关。
   *
   * 为 false 时（手动录入、来源与投递链接都指不到站点的岗位）勾选框禁用、也不计入
   * 「待投递」——投递台只能驱动招聘网站上的岗位，强投只会得到一条失败记录。
   *
   * **可选**：缺失按"支持"处理（与后端默认同向），判断写 `=== false`，别写 `!x`。
   */
  apply_supported?: boolean;
  created_at: string;
  updated_at: string;
}

export interface ApplyQueueAddItem {
  job_id: number;
  resume_id?: number | null;
  greeting?: string;
  confirm_real_gap?: boolean;
  confirm_unanalyzed?: boolean;
}

/** 入队被 409 拦下时后端回传的结构化原因（用于逐条确认）。 */
export interface QueueConflictDetail {
  message: string;
  job_id?: number;
  gaps?: string[];
  unanalyzed?: boolean;
  /**
   * 岗位来源不是投递台支持的招聘网站。
   *
   * 这一类比 `unanalyzed` / `gaps` 更硬：那两类用户可以显式确认后放行，这一类**确认也没用**
   * ——投递时根本定位不到站点，所以界面只提示、不给"仍然加入"的按钮。
   */
  site_unsupported?: boolean;
  /** 被拦下的岗位（开始投递时可能不止一个）。 */
  job_ids?: number[];
  job_titles?: string[];
}

// ===== ⑧ 批次与记录 =====
export interface ApplyTask {
  id: number;
  kind: TaskKind;
  status: TaskStatus;
  total: number;
  processed: number;
  succeeded: number;
  failed: number;
  skipped: number;
  current_step: string;
  stop_reason: string;
  config: Record<string, unknown>;
  message: string;
  started_at: string | null;
  finished_at: string | null;
  created_at: string;
}

export interface ApplyTaskItem {
  id: number;
  task_id: number;
  job_id: number | null;
  job_title: string;
  company: string;
  resume_id: number | null;
  resume_title: string;
  greeting: string;
  status: TaskItemStatus;
  failure_category: string;
  failure_detail: string;
  attempt: number;
  sort_order: number;
  started_at: string | null;
  finished_at: string | null;
  created_at: string;
}

export interface ApplyTaskDetail extends ApplyTask {
  items: ApplyTaskItem[];
}

export interface ApplyRecord {
  id: number;
  task_id: number;
  job_id: number | null;
  job_title: string;
  company: string;
  resume_title: string;
  greeting: string;
  status: TaskItemStatus;
  failure_category: string;
  failure_label: string;
  failure_detail: string;
  attempt: number;
  created_at: string;
  finished_at: string | null;
}

/**
 * 一个投递批次及其全部记录：投递记录按批次分组展示的载体。
 *
 * 一次「开始投递」= 一个批次；用户一次性投 N 个岗位时，这 N 条记录同属一组。
 * 组头上的统计是该批次自己的账（不随筛选变化），组内是命中的记录。
 */
export interface ApplyRecordBatch {
  id: number;
  status: TaskStatus;
  total: number;
  processed: number;
  succeeded: number;
  failed: number;
  skipped: number;
  message: string;
  created_at: string;
  finished_at: string | null;
  items: ApplyRecord[];
}

export interface GreetingPreview {
  greeting: string;
  source: "generated" | "queue" | "default";
}

// ===== ⑫ 站点健康度（degraded 标记）=====

/**
 * 站点健康度状态：`ok` 正常；`degraded` 疑似改版（采集悄悄抓不到东西）。
 *
 * 判据完全在后端（`app/services/site_health.py`）。前端**只读** `status` 与 `reasons`，
 * 绝不在这里再判一次——否则两处判据迟早漂移。
 */
export type SiteHealthStatus = "ok" | "degraded";

/** 一次采集运行的摘要（供界面展开对照「采集记录」）。 */
export interface CollectRunSummary {
  status: string;
  failure_category: string;
  succeeded: number;
  detail_missing: number;
  created_at: string;
}

export interface SiteHealth {
  site_key: string;
  display_name: string;
  status: SiteHealthStatus;
  /** 人类可读、可操作的中文原因（来自后端）。degraded 时非空。 */
  reasons: string[];
  sampled: number;
  selector_failures: number;
  detail_drift_runs: number;
  recent: CollectRunSummary[];
}

export interface SiteHealthList {
  sites: SiteHealth[];
}

/** 取当前站点的健康度；找不到时回退 undefined（前端不自行判定，只如实展示后端结论）。 */
export function siteHealthFor(
  list: SiteHealthList | undefined,
  siteKey: string,
): SiteHealth | undefined {
  return list?.sites.find((site) => site.site_key === siteKey);
}

File: backend/app/models/apply.py
"""自动投递中心的数据模型：匹配结论、投递队列与执行批次。

本模块把"采集进来 → 判断够不够 → 显式点头 → 自动投出去"这条链路里需要**持久化**
的部分沉淀成 4 张表：

- ``job_match_analysis``：一个岗位**保留最近一次**匹配结论（重复分析即覆盖），供准入
  判断与界面展示；不建到资料/简历的外键——分析只读资料，证据以文本形式存在结果里。
- ``apply_queue_item``：用户显式勾选的待投递队列。同一岗位在队列里只能有一条。
- ``apply_task``：一次执行批次（采集与投递共用同一张表，靠 ``kind`` 区分）。
- ``apply_task_item``：批次内每个岗位的执行条目，也就是投递记录。

时间戳统一用 ``.profile.utcnow``（无时区 UTC）；岗位/简历被删除后记录仍要可读，所以
对它们用 ``SET NULL`` + 快照字段；批次内的条目随批次 ``CASCADE`` 删除。

本模块同时定义前后端共用的**状态/枚举常量**与唯一的准入闸门映射 ``admission_of``。
只在这里写一次，前端镜像同一份取值，避免"两处各判一次、判得不一样"。
"""
from datetime import datetime
from typing import Any

from sqlalchemy import JSON, Boolean, DateTime, ForeignKey, Integer, String, Text, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column

from ..database import Base
from .profile import utcnow

# ===== ① 匹配状态（五类，前后端共用一份取值）=====
MATCH_STATUS_MATCHED = "matched"
MATCH_STATUS_EXPRESSION_GAP = "expression_gap"
MATCH_STATUS_EVIDENCE_INSUFFICIENT = "evidence_insufficient"
MATCH_STATUS_REAL_GAP = "real_gap"
MATCH_STATUS_TO_CONFIRM = "to_confirm"
MATCH_STATUSES = (
    MATCH_STATUS_MATCHED,
    MATCH_STATUS_EXPRESSION_GAP,
    MATCH_STATUS_EVIDENCE_INSUFFICIENT,
    MATCH_STATUS_REAL_GAP,
    MATCH_STATUS_TO_CONFIRM,
)

# ===== ② 准入闸门（唯一权威：只在这里映射一次）=====
ADMISSION_ALLOW = "allow"
ADMISSION_BLOCK = "block"
ADMISSION_NEEDS_CONFIRM = "needs_confirm"
ADMISSIONS = (ADMISSION_ALLOW, ADMISSION_BLOCK, ADMISSION_NEEDS_CONFIRM)

# 硬性门槛结论（冗余在 job_match_analysis.hard_gate，便于查询与展示）。
HARD_GATE_MET = "met"
HARD_GATE_UNMET = "unmet"
HARD_GATE_UNKNOWN = "unknown"
HARD_GATES = (HARD_GATE_MET, HARD_GATE_UNMET, HARD_GATE_UNKNOWN)

_ALLOW_STATUSES = frozenset({MATCH_STATUS_MATCHED, MATCH_STATUS_EXPRESSION_GAP})
_BLOCK_STATUSES = frozenset({MATCH_STATUS_REAL_GAP})
_NEEDS_CONFIRM_STATUSES = frozenset(
    {MATCH_STATUS_EVIDENCE_INSUFFICIENT, MATCH_STATUS_TO_CONFIRM}
)


def admission_of(status: str) -> str:
    """把单条匹配状态映射到投递准入结论。

    ``已匹配`` / ``表达缺口`` → 允许自动投；``真实缺口`` → 不投；
    ``证据不足`` / ``待确认`` → 需用户逐条确认。未知取值按"需确认"处理（保守兜底，
    绝不因为读到一个没见过的状态就放行自动投递）。
    """
    if status in _ALLOW_STATUSES:
        return ADMISSION_ALLOW
    if status in _BLOCK_STATUSES:
        return ADMISSION_BLOCK
    if status in _NEEDS_CONFIRM_STATUSES:
        return ADMISSION_NEEDS_CONFIRM
    return ADMISSION_NEEDS_CONFIRM


def requires_confirmation(status: str) -> bool:
    """该匹配状态是否属于"需用户逐条确认"一类。"""
    return admission_of(status) == ADMISSION_NEEDS_CONFIRM


# ===== ③ 队列条目状态 =====
QUEUE_STATUS_PENDING = "pending"
QUEUE_STATUS_SKIPPED = "skipped"
QUEUE_STATUS_DONE = "done"
QUEUE_STATUSES = (QUEUE_STATUS_PENDING, QUEUE_STATUS_SKIPPED, QUEUE_STATUS_DONE)

# ===== ④ 批次种类 =====
TASK_KIND_COLLECT = "collect"
TASK_KIND_APPLY = "apply"
TASK_KINDS = (TASK_KIND_COLLECT, TASK_KIND_APPLY)

# ===== ⑤ 批次状态机 =====
TASK_STATUS_PENDING = "pending"
TASK_STATUS_RUNNING = "running"
TASK_STATUS_PAUSED = "paused"
TASK_STATUS_BREAKER_PAUSED = "breaker_paused"
TASK_STATUS_COMPLETED = "completed"
TASK_STATUS_STOPPED = "stopped"
TASK_STATUS_FAILED = "failed"
TASK_STATUSES = (
    TASK_STATUS_PENDING,
    TASK_STATUS_RUNNING,
    TASK_STATUS_PAUSED,
    TASK_STATUS_BREAKER_PAUSED,
    TASK_STATUS_COMPLETED,
    TASK_STATUS_STOPPED,
    TASK_STATUS_FAILED,
)

# ===== ⑥ 批次条目状态机 =====
ITEM_STATUS_PENDING = "pending"
ITEM_STATUS_RUNNING = "running"
ITEM_STATUS_SUCCESS = "success"
ITEM_STATUS_FAILED = "failed"
ITEM_STATUS_SKIPPED = "skipped"
ITEM_STATUSES = (
    ITEM_STATUS_PENDING,
    ITEM_STATUS_RUNNING,
    ITEM_STATUS_SUCCESS,
    ITEM_STATUS_FAILED,
    ITEM_STATUS_SKIPPED,
)

# ===== ⑦ 失败分类（一等的可识别类型）=====
FAILURE_SELECTOR_INVALID = "selector_invalid"
FAILURE_LOGIN_REQUIRED = "login_required"
FAILURE_CAPTCHA_REQUIRED = "captcha_required"
FAILURE_GREETING_MISSING = "greeting_missing"
FAILURE_NETWORK_TIMEOUT = "network_timeout"
FAILURE_FILE_UPLOAD_FAILED = "file_upload_failed"
# 岗位不属于任何已注册的招聘网站（例如纯手动录入、来源与投递链接都指不到站点）。
# 它**不是**运行时故障，而是"这个岗位本来就不该走到投递这一步"——所以单列一类，
# 否则记录里只会写「未知失败」，用户与我们都看不出真正的原因。
FAILURE_SITE_UNSUPPORTED = "site_unsupported"
FAILURE_UNKNOWN = "unknown"
FAILURE_CATEGORIES = (
    FAILURE_SELECTOR_INVALID,
    FAILURE_LOGIN_REQUIRED,
    FAILURE_CAPTCHA_REQUIRED,
    FAILURE_GREETING_MISSING,
    FAILURE_NETWORK_TIMEOUT,
    FAILURE_FILE_UPLOAD_FAILED,
    FAILURE_SITE_UNSUPPORTED,
    FAILURE_UNKNOWN,
)

# 面向用户的失败分类说明：诊断信息会直接展示给用户，用户再反馈给我们修，所以要写清楚。
FAILURE_CATEGORY_LABELS = {
    FAILURE_SELECTOR_INVALID: "页面结构变化 / 选择器失效",
    FAILURE_LOGIN_REQUIRED: "需要登录",
    FAILURE_CAPTCHA_REQUIRED: "需要验证码或安全验证",
    FAILURE_GREETING_MISSING: "招呼语缺失",
    FAILURE_NETWORK_TIMEOUT: "网络超时",
    FAILURE_FILE_UPLOAD_FAILED: "简历上传失败",
    FAILURE_SITE_UNSUPPORTED: "岗位来源不支持自动投递",
    FAILURE_UNKNOWN: "未知失败",
}

# ===== ⑧ 执行步骤（apply_task.current_step）=====
STEP_OPENING = "opening"
STEP_FILLING = "filling"
STEP_GREETING = "greeting"
STEP_UPLOADING = "uploading"
STEP_SUBMITTING = "submitting"
STEP_VERIFYING = "verifying"
STEP_WAITING = "waiting"
STEP_IDLE = "idle"
TASK_STEPS = (
    STEP_OPENING,
    STEP_FILLING,
    STEP_GREETING,
    STEP_UPLOADING,
    STEP_SUBMITTING,
    STEP_VERIFYING,
    STEP_WAITING,
    STEP_IDLE,
)

# ===== ⑨ 停止原因 =====
STOP_REASON_USER = "user"
STOP_REASON_BREAKER = "breaker"
STOP_REASON_DONE = "done"
STOP_REASON_ERROR = "error"
STOP_REASONS = (STOP_REASON_USER, STOP_REASON_BREAKER, STOP_REASON_DONE, STOP_REASON_ERROR)


class JobMatchAnalysis(Base):
    """岗位匹配度分析结果：一个岗位保留最近一次，供准入与展示。"""

    __tablename__ = "job_match_analysis"

    id: Mapped[int] = mapped_column(primary_key=True)
    # 一个岗位只保留最近一次分析（job_id 唯一）：重复分析即覆盖。岗位删除后置空，
    # 但 job_title / company 快照与结果仍在，历史结论不会凭空消失。
    job_id: Mapped[int | None] = mapped_column(
        ForeignKey("job.id", ondelete="SET NULL"), nullable=True, index=True, unique=True
    )
    job_title: Mapped[str] = mapped_column(String(128), default="")
    company: Mapped[str] = mapped_column(String(128), default="")
    # 完整匹配结果（结构见设计 §3.4）：硬性条件、核心能力、加分项、结论、建议、口径说明。
    result: Mapped[dict[str, Any]] = mapped_column(JSON, default=dict)
    # 硬性门槛结论：met / unmet / unknown（准入闸门字段，单独冗余便于查询）。
    hard_gate: Mapped[str] = mapped_column(String(16), default=HARD_GATE_UNKNOWN, index=True)
    # 是否含需逐条确认项（待确认 / 证据不足）。
    requires_confirm: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    model: Mapped[str] = mapped_column(String(64), default="")
    created_at: Mapped[datetime] = mapped_column(DateTime, default=utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=utcnow, onupdate=utcnow)


class ApplyQueueItem(Base):
    """投递队列条目：用户显式勾选、按顺序待投递的岗位。"""

    __tablename__ = "apply_queue_item"
    __table_args__ = (UniqueConstraint("job_id", name="uq_apply_queue_item_job_id"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    # 队列条目只来自"岗位广场里已存在的岗位"；岗位被删除后条目仍可读（快照 + SET NULL）。
    job_id: Mapped[int | None] = mapped_column(
        ForeignKey("job.id", ondelete="SET NULL"), nullable=True, index=True
    )
    job_title: Mapped[str] = mapped_column(String(128), default="")
    company: Mapped[str] = mapped_column(String(128), default="")
    # 用户指定用哪份简历；空 = 运行时按默认规则解析。
    resume_id: Mapped[int | None] = mapped_column(
        ForeignKey("resume_record.id", ondelete="SET NULL"), nullable=True
    )
    # 本岗位的招呼语（用户可编辑 / 按岗位生成后落此）；空 = 用默认招呼语。
    greeting: Mapped[str] = mapped_column(Text, default="")
    sort_order: Mapped[int] = mapped_column(Integer, default=0, index=True)
    # pending(待投递) / skipped(用户移出前保留) / done(已投递)。
    status: Mapped[str] = mapped_column(String(16), default=QUEUE_STATUS_PENDING, index=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=utcnow, onupdate=utcnow)


class ApplyTask(Base):
    """一次执行批次：采集与投递共用，靠 kind 区分。"""

    __tablename__ = "apply_task"

    id: Mapped[int] = mapped_column(primary_key=True)
    # collect(采集) / apply(投递)。
    kind: Mapped[str] = mapped_column(String(16), index=True)
    # 状态机取值见模块顶部常量（pending/running/paused/breaker_paused/completed/...）。
    status: Mapped[str] = mapped_column(String(16), default=TASK_STATUS_PENDING, index=True)
    # 数量化进度（**不用百分比**）：总数与各状态计数。
    total: Mapped[int] = mapped_column(Integer, default=0)
    processed: Mapped[int] = mapped_column(Integer, default=0)
    succeeded: Mapped[int] = mapped_column(Integer, default=0)
    failed: Mapped[int] = mapped_column(Integer, default=0)
    skipped: Mapped[int] = mapped_column(Integer, default=0)
    # 当前步骤（opening/filling/greeting/submitting/waiting/...）。
    current_step: Mapped[str] = mapped_column(String(32), default="")
    # 停止原因：user/breaker/done/error。
    stop_reason: Mapped[str] = mapped_column(String(32), default="")
    # 本次运行生效的配置快照（间隔/上限/熔断阈值…），留痕便于复现。
    config: Mapped[dict[str, Any]] = mapped_column(JSON, default=dict)
    # 醒目提示文案（熔断、需登录等）。
    message: Mapped[str] = mapped_column(Text, default="")
    started_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    finished_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=utcnow)


class ApplyTaskItem(Base):
    """批次内每岗位执行条目，也就是一条投递记录。"""

    __tablename__ = "apply_task_item"

    id: Mapped[int] = mapped_column(primary_key=True)
    # 条目随批次删除（CASCADE）。
    task_id: Mapped[int] = mapped_column(
        ForeignKey("apply_task.id", ondelete="CASCADE"), index=True
    )
    job_id: Mapped[int | None] = mapped_column(
        ForeignKey("job.id", ondelete="SET NULL"), nullable=True, index=True
    )
    job_title: Mapped[str] = mapped_column(String(128), default="")
    company: Mapped[str] = mapped_column(String(128), default="")
    # 快照：记录当时用的简历，简历被删除后仍可读。
    resume_id: Mapped[int | None] = mapped_column(
        ForeignKey("resume_record.id", ondelete="SET NULL"), nullable=True
    )
    resume_title: Mapped[str] = mapped_column(String(256), default="")
    # 实际发送的招呼语（脱敏按"记全文但截断到合理上限"处理）。
    greeting: Mapped[str] = mapped_column(Text, default="")
    # pending / running / success / failed / skipped。
    status: Mapped[str] = mapped_column(String(16), default=ITEM_STATUS_PENDING, index=True)
    # 失败分类（selector_invalid / captcha_required / ...），空串表示未失败。
    failure_category: Mapped[str] = mapped_column(String(32), default="")
    # 可操作诊断（当前 URL / 页面标题 / 匹配控件数 / 期望控件描述）。
    failure_detail: Mapped[str] = mapped_column(Text, default="")
    attempt: Mapped[int] = mapped_column(Integer, default=0)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    started_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    finished_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=utcnow)
